opentreemap/treemap/udf.py

```diff
@@ -766,111 +877,115 @@ def datatype_by_field(self):
             try:
                 return float(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s '
-                                        'must be a real number') %
-                                      {'fieldname': self.name})
+                raise complaint('%(fieldname)s must be a real number',
+                                fieldname=self.name)
         elif datatype == 'int':
             try:
                 if float(value) != int(value):
                     raise ValueError
 
                 return int(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s must be an integer') %
-                                      {'fieldname': self.name})
-        elif datatype == 'user':
-            if isinstance(value, User):
-                return value
-
-            try:
-                pk = int(value)
-            except ValueError:
-                raise ValidationError(_('%(fieldname)s must be an integer') %
-                                      {'fieldname': self.name})
-            try:
-                return User.objects.get(pk=pk)
-            except User.DoesNotExist:
-                raise ValidationError(_('%(fieldname)s user not found') %
-                                      {'fieldname': self.name})
-
+                raise complaint('%(fieldname)s must be an integer',
+                                fieldname=self.name)
         elif datatype == 'date':
             if isinstance(value, (date, datetime)):
                 return value
 
             try:
                 valid_date = parse_date_string_with_or_without_time(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s must be formatted as '
-                                        'YYYY-MM-DD') %
-                                      {'fieldname': self.name})
+                raise complaint(('%(fieldname)s must be formatted as '
+                                 'YYYY-MM-DD'),
+                                fieldname=self.name)
 
             # Ensure date UDF values contain a year >= 1900 so that
             # date formatting with `strftime` will work correctly.
             if valid_date.year < 1900:
-                raise ValidationError(_('%(fieldname)s year must be >= 1900')
-                                      % {'fieldname': self.name})
+                raise complaint('%(fieldname)s year must be >= 1900',
+                                fieldname=self.name)
 
             return valid_date
 
         elif 'choices' in datatype_dict:
             def _validate(val):
                 if val not in datatype_dict['choices']:
-                    raise ValidationError(
-                        _('Invalid choice (%(given)s). Expecting %(allowed)s')
-                        % {'given': val,
-                           'allowed': ', '.join(datatype_dict['choices'])})
+                    raise complaint(
```

Comment:
I verified that I could make a new instance with a choice field on Tree, set a choice value on a tree, then clear the choice value and save successfully.

However, I ran into an exception running through these steps on a local dev copy of treemapla

- Browse http://localhost:6060/latreemap/features/2981708/edit
- Click save
- Click save, Generic "oops" error toast is shown. The AJAX PUT returns this exception

```
Traceback:
File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  46.             return view_fn(request, instance, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in routed
  35.             return kwargs['ELSE'](request, *args2, **kwargs2)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  196.             return view_fn(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in newreq
  125.                 outp = req_function(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in newreq
  110.             return '%s' % json.dumps(outp, cls=LazyEncoder)
File "/usr/lib/python2.7/json/__init__.py" in dumps
  250.         sort_keys=sort_keys, **kw).encode(obj)
File "/usr/lib/python2.7/json/encoder.py" in encode
  207.         chunks = self.iterencode(o, _one_shot=True)
File "/usr/lib/python2.7/json/encoder.py" in iterencode
  270.         return _iterencode(o, 0)
File "/opt/app/core/opentreemap/treemap/udf.py" in __getitem__
  1206.                 return udf.clean_value(value)
File "/opt/app/core/opentreemap/treemap/udf.py" in clean_value
  954.                 _validate(value)
File "/opt/app/core/opentreemap/treemap/udf.py" in _validate
  951.                         allowed=', '.join(datatype_dict['choices']))

Exception Type: ValidationError at /latreemap/features/2981708/
Exception Value: {u'udf:Funded by': [u'Invalid choice (None). Expecting CAL FIRE, CalTrans, City Plants, SoCal Edison, Adjacent property owner']}
```

- Refresh the page. Crash on a template error.

```
Template error:
In template /opt/app/core/opentreemap/treemap/templates/treemap/partials/tree_section.html, error at line 20
   {u'udf:Funded by': u'Invalid choice (None). Expecting CAL FIRE, CalTrans, City Plants, SoCal Edison, Adjacent property owner'}

   10 : <!-- Add a Tree Section -->



   11 : {% include "treemap/partials/plot_detail_add_tree.html" %}



   12 : 



   13 : <!-- Tree Information -->



   14 : <div id="tree-details" {{ has_tree|yesno:",style=display:none;" }}>



   15 :     <table class="table table-hover">



   16 :         <tbody>



   17 :         {% for tuple in group.fields %}



   18 :             {% with field=tuple.0 label=tuple.1 template=tuple.2 %}



   19 :                 {% if label %}



   20 :                      {% field label from field for request.user withtemplate template %} 



   21 :                 {% else %}



   22 :                     {% field from field for request.user withtemplate template %}



   23 :                 {% endif %}



   24 :             {% endwith %}



   25 :         {% endfor %}



   26 : {#         {% usercanread tree "id" as value %} #}



   27 : {#         {% endusercanread %} #}



   28 :         </tbody>



   29 :     </table>



   30 : 


Traceback:
File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  46.             return view_fn(request, instance, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in routed
  33.             return req_method(request, *args2, **kwargs2)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in render_map_feature_detail
  159.                               **kwargs)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in map_feature_detail
  94.         return render_to_response(template, context, RequestContext(request))
File "/usr/local/lib/python2.7/dist-packages/django/shortcuts.py" in render_to_response
  45.             using=using)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader.py" in render_to_string
  115.                         template_name, context, context_instance, dirs, dictionary)
File "/usr/local/lib/python2.7/dist-packages/django/template/engine.py" in render_to_string
  221.             return t.render(context_instance)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  210.                     return self._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  135.         return compiled_parent._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  135.         return compiled_parent._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  65.                 result = block.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  159.                 return template.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  212.                 return self._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  135.         return compiled_parent._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  65.                 result = block.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  217.                             nodelist.append(node.render(context))
File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  329.                 return nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  159.                 return template.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  212.                 return self._render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  202.         return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  217.                             nodelist.append(node.render(context))
File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  576.             return self.nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  329.                 return nodelist.render(context)
File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  905.                 bit = self.render_node(node, context)
File "/usr/local/lib/python2.7/dist-packages/django/template/debug.py" in render_node
  79.             return node.render(context)
File "/opt/app/core/opentreemap/treemap/templatetags/form_extras.py" in render
  382.             field_value = _field_value(model, field_name, data_type)
File "/opt/app/core/opentreemap/treemap/templatetags/form_extras.py" in _field_value
  360.                     val = model.udfs[udf_field_name]
File "/opt/app/core/opentreemap/treemap/udf.py" in __getitem__
  1206.                 return udf.clean_value(value)
File "/opt/app/core/opentreemap/treemap/udf.py" in clean_value
  954.                 _validate(value)
File "/opt/app/core/opentreemap/treemap/udf.py" in _validate
  951.                         allowed=', '.join(datatype_dict['choices']))

Exception Type: ValidationError at /latreemap/features/2978539/
Exception Value: {u'udf:Funded by': [u'Invalid choice (None). Expecting CAL FIRE, CalTrans, City Plants, SoCal Edison, Adjacent property owner']}
```